{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " o sinal total é 10217360.0\n",
      "[' ([M+H]+)', '([M+Na]+)']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "BYallmet = pd.read_csv('KOG1.tsv', sep = '\\t')#.drop_duplicates(subset='KEGG_cid',keep='first')\n",
    "\n",
    "BYallmet = BYallmet.set_index('KEGG_cid')\n",
    "\n",
    "height = BYallmet.peak_height.C00127.sum()\n",
    "\n",
    "mass = BYallmet['KEGG_name'].C00127\n",
    "\n",
    "last = []\n",
    "for n in mass:\n",
    "    last.append(n[-9:])\n",
    "\n",
    "print(f' o sinal total é {height}')\n",
    "\n",
    "print(last)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64\n"
     ]
    }
   ],
   "source": [
    "def readFASTA(filename):\n",
    "    \"\"\"This function reads a FASTA format file and\n",
    "    returns a pair of strings\n",
    "    with the header and the sequence\n",
    "    \"\"\"\n",
    "    with open(filename) as a:\n",
    "        lines = [line.strip() for line in a]\n",
    "    lines = [line for line in lines if len(line) > 0]\n",
    "    \n",
    "    if lines[0].startswith('>'):\n",
    "        return lines[0], ''.join(lines[1:])\n",
    "    else:\n",
    "        return '', ''.join(lines)\n",
    "\n",
    "# A função 'open' lê cada linha como um elemento individual do ficheiro.\n",
    "# É quase como se cada linha fosse um elemento da lista gerada com a função.\n",
    "# A função retorna uma string de acordo com as operações mencionadas.\n",
    "\n",
    "basesDNA = 'ATGC'\n",
    "basesRNA = 'AUGC'\n",
    "\n",
    "aa_residues   = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "\n",
    "complementDNA = { 'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "complementRNA = { 'A':'U', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "gencode = {'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L', 'UCU': 'S',\n",
    "     'UCC': 'S', 'UCA': 'S', 'UCG': 'S', 'UAU': 'Y', 'UAC': 'Y',\n",
    "     'UGU': 'C', 'UGC': 'C', 'UGG': 'W', 'CUU': 'L', 'CUC': 'L',\n",
    "     'CUA': 'L', 'CUG': 'L', 'CCU': 'P', 'CCC': 'P', 'CCA': 'P',\n",
    "     'CCG': 'P', 'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "     'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AUU': 'I',\n",
    "     'AUC': 'I', 'AUA': 'I', 'AUG': 'M', 'ACU': 'T', 'ACC': 'T',\n",
    "     'ACA': 'T', 'ACG': 'T', 'AAU': 'N', 'AAC': 'N', 'AAA': 'K',\n",
    "     'AAG': 'K', 'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "     'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V', 'GCU': 'A',\n",
    "     'GCC': 'A', 'GCA': 'A', 'GCG': 'A', 'GAU': 'D', 'GAC': 'D',\n",
    "     'GAA': 'E', 'GAG': 'E', 'GGU': 'G', 'GGC': 'G', 'GGA': 'G',\n",
    "     'GGG': 'G', 'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP'}\n",
    "\n",
    "trans123 = {'A': 'Ala', 'C': 'Cys', 'E': 'Glu', 'D': 'Asp', 'G': 'Gly',\n",
    "            'F': 'Phe', 'I': 'Ile', 'H': 'His', 'K': 'Lys', 'M': 'Met',\n",
    "            'L': 'Leu', 'N': 'Asn', 'Q': 'Gln', 'P': 'Pro', 'S': 'Ser',\n",
    "            'R': 'Arg', 'T': 'Thr', 'W': 'Trp', 'V': 'Val', 'Y': 'Tyr'}\n",
    "\n",
    "aa_masses = {'A': 71.0788, \n",
    "             'C': 103.1448, \n",
    "             'E': 129.1155, \n",
    "             'D': 115.0886, \n",
    "             'G': 57.0519, \n",
    "             'F': 147.1766, \n",
    "             'I': 113.1594, \n",
    "             'H': 137.1411, \n",
    "             'K': 128.1741, \n",
    "             'M': 131.1926, \n",
    "             'L': 113.1594, \n",
    "             'N': 114.1038, \n",
    "             'Q': 128.1307, \n",
    "             'P': 97.11667, \n",
    "             'S': 87.0782, \n",
    "             'R': 156.1875, \n",
    "             'T': 101.1051, \n",
    "             'W': 186.2132, \n",
    "             'V': 99.1326, \n",
    "             'Y': 163.176\n",
    "             }\n",
    "\n",
    "aa_classes = { 'small'       : 'PCAGVTDSN',\n",
    "    'tiny'        : 'AGCS',\n",
    "    'aliphatic'   : 'ILV',\n",
    "    'aromatic'    : 'FYWH',\n",
    "    'positive'    : 'KHR',\n",
    "    'negative'    : 'DE',\n",
    "    'charged'     : 'KHRDE',\n",
    "    'hydrophobic' : 'CAGIVLTMHYWF',\n",
    "    'polar'       : \"CSTNDYWHKQRDE\",\n",
    "    'proline'     : 'P'}\n",
    "\n",
    "print(len(gencode))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCAGTTCCTAGGTCGTTCCT appears in the sequence complementary to the requested one\n",
      "AAGGGGCTTTACGATGGAGT appears in the requested sequence\n",
      "CGATCAGGCGAAAATAGC appears in the requested sequence\n",
      "TACGCGGTGTTTTTTAGTTATA appears in the sequence complementary to the requested one\n",
      "The requested amplicon has 336 bp\n",
      "The requested amplicon has 489 bp\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CGATCAGGCGAAAATAGCATTCTTTCAATGTATCAAAAATGATTCCAGAAACAGCGAAAAATTGGTTGATATATTGTGTCAAAGGCTAAATTTCAATTTTGATAAGAGAGCCAAAGAAATGAATTTAGATCTAAGAGTACATTCATTAGATGTTGAAGATTACATCGAGCTGACTGATAACAAGGAATTCAAAAACTTGATATCATCCGGCGTTGAAAAAGTGACAAGATCTCAAAAAGATTTGTTTACGCTAAAATACAAGAGAGTGCAGAGAATTGTGCCTCATAACGATACTTTGATTGAATTATTCGATCAAGATATTGTTATGCATATGAGTGAATTGCAATTGGTATTAACACCTAGATCTGTATTAACTTTAATGAATTATGCAATGCTTACTTTTACGGATCCTAACGCTCCAGAAATGCCGGCTGATGTGTTGAGACATAATAAAGAAGATAGGGACGATGCGCCACAAAAAATCAATAT'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = readFASTA('GLO1_flanking.fsa')[1]\n",
    "gene2 = readFASTA('VPS13_flanking.fsa')[1]\n",
    "\n",
    "complementair = []\n",
    "for n in gene:\n",
    "    complementair.append(complementDNA[n])\n",
    "comp = ''.join(complementair)\n",
    "\n",
    "#GLO1:\n",
    "primerfw = 'AAGGGGCTTTACGATGGAGT' #5' - 3'\n",
    "primerrv = 'TCAGTTCCTAGGTCGTTCCT' #3' - 5'\n",
    "\n",
    "#VPS13:\n",
    "primerfw1 = 'CGATCAGGCGAAAATAGC'\n",
    "primerrv1 = 'TACGCGGTGTTTTTTAGTTATA'\n",
    "def complement(seq):\n",
    "    complementair = []\n",
    "    for n in seq:\n",
    "        complementair.append(complementDNA[n])\n",
    "    comp = ''.join(complementair)\n",
    "    return comp\n",
    "\n",
    "def prot_noncod(seq): #From the noncoding chain it produces the associated protein\n",
    "    comp = complement(seq)\n",
    "    mRNA = []\n",
    "    for n in comp:\n",
    "        mRNA.append(complementRNA[n])\n",
    "    mRNA = ''.join(mRNA)\n",
    "\n",
    "    protein = []\n",
    "    cod = [mRNA[i] + mRNA[i+1] + mRNA[i+2] for i in range(0, len(mRNA), 3)]\n",
    "    for n in cod:\n",
    "        protein.append(gencode[n])\n",
    "    protein = '-'.join(protein)\n",
    "    return protein\n",
    "\n",
    "def prot_cod(seq): #From the noncoding chain it produces the associated protein\n",
    "    mRNA = []\n",
    "    for n in seq:\n",
    "        mRNA.append(complementRNA[n])\n",
    "    mRNA = ''.join(mRNA)\n",
    "\n",
    "    protein = []\n",
    "    cod = [mRNA[i] + mRNA[i+1] + mRNA[i+2] for i in range(0, len(mRNA), 3)]\n",
    "    for n in cod:\n",
    "        protein.append(gencode[n])\n",
    "    protein = '-'.join(protein)\n",
    "    return protein\n",
    "    \n",
    "def BLAST(seq, seq1): #For a sequence (seq), this function analyses if another, smaller sequence (seq1) is contained inside the first one\n",
    "    complementair = []\n",
    "    for n in seq:\n",
    "        complementair.append(complementDNA[n])\n",
    "    comp = ''.join(complementair)\n",
    "\n",
    "    anneal = str(seq1)\n",
    "\n",
    "    if anneal in seq:\n",
    "        return print(f'{anneal} appears in the requested sequence')\n",
    "    elif anneal in comp:\n",
    "        return print(f'{anneal} appears in the sequence complementary to the requested one')\n",
    "    else:\n",
    "        return print(f'{anneal} does not appear in any sequence') \n",
    "\n",
    "def gene_size(gene, primerfw, primerrv):\n",
    "    if primerfw in gene:\n",
    "        a = gene.split(primerfw)[1]\n",
    "        a = primerfw + a\n",
    "        a = a.split(complement(primerrv))[0]\n",
    "        a = a + complement(primerrv)\n",
    "        size = len(a)\n",
    "    return print(f'The requested amplicon has {size} bp')\n",
    "\n",
    "def amplicon(gene, primerfw, primerrv):\n",
    "    if primerfw in gene:\n",
    "        a = gene.split(primerfw)[1]\n",
    "        a = primerfw + a\n",
    "        a = a.split(complement(primerrv))[0]\n",
    "        a = a + complement(primerrv)\n",
    "    return a\n",
    "\n",
    "BLAST(gene, primerrv)\n",
    "BLAST(gene, primerfw)\n",
    "BLAST(gene2, primerfw1)\n",
    "BLAST(gene2, primerrv1)\n",
    "gene_size(gene, primerfw, primerrv)\n",
    "gene_size(gene2, primerfw1, primerrv1)\n",
    "amplicon(gene2, primerfw1, primerrv1)\n",
    "\n",
    "\n",
    "#print(prot_noncod(gene))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1c0e22491d37f29e12dbbed1764984b5047c336fcb37effe27c9c012a2ac9d2b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
